### PR TITLE
Lazy load room members - Part I

### PR DESCRIPTION
**Pull request by bwindels**
Lazy loading of members, for now using the /joined_members endpoint until we have something better.
 - [x] prototype lazily loaded members with /joined_members
- [x] pull code that assumes `RoomMember.events...` into RoomMember so it can transparently take lazily loaded members into account. 
 - [x] return fake sentinels with just user id when users haven't been loaded yet, better than breaking message continuation.
 - [x] timeline reset should take into account lazily loaded members of previous timeline
 - [x] use `/members` endpoint
 - [x] fix rooms not visible on initial sync where user wasn't active recently 
- [x] write/fix unit tests

Parent issue: https://github.com/vector-im/riot-web/issues/6611

### CODE REVIEW

File: src/models/room-state.js [hunk 1]
```diff
@@ -287,7 +321,7 @@ RoomState.prototype.getLastModifiedTime = function() {
 /**
  * Get user IDs with the specified display name.
  * @param {string} displayName The display name to get user IDs from.
- * @return {string[]} An array of user IDs or an empty array.
+ * @return {string[]} An array of user IDs or an empty array.
```
Comment (t3chguy) [inline src/models/room-state.js]: seems like a spurious trailing whitespace was added for no reason?

**Comment by bwindels** [comment]
want to do a general review if it's ready to merge, but all should be here for a first PR.
Note that the react SDK code won't call into lazy loading code if the feature is not enabled.

File: src/models/room.js [hunk 2]
```diff
@@ -201,6 +204,27 @@ Room.prototype.getLiveTimeline = function() {
...
+    return this._membersNeedLoading;
+};
+
+/**
+ * Sets the lazily loaded members from the result of calling /joined_members
+ * @param {Promise} joinedMembersPromise promise with result of /joined_members endpoint
+ */
+Room.prototype.setLazilyLoadedMembers = async function(joinedMembersPromise) {
+    this._membersNeedLoading = false;
+    const members = await joinedMembersPromise;
```
Comment (krombel) [inline src/models/room.js]: You might want to catch the case that this promise fails - e.g. when something is weird with the connection or server

File: src/models/room.js [hunk 3]
```diff
@@ -201,6 +204,27 @@ Room.prototype.getLiveTimeline = function() {
...
+    //wait 10 seconds
+    await new Promise((resolve) => setTimeout(resolve, 5000));
```
Comment (krombel) [inline src/models/room.js]: Why should this wait? `await joinedMembersPromise` should make sure that it only continues executing once it is (successfully) done.
Besides that "5000" != 10 seconds ;)

File: src/models/room.js (hunk 2 above)
Comment (bwindels) [reply src/models/room.js]: Good point!

File: src/client.js [hunk 4]
```diff
@@ -742,6 +742,20 @@ MatrixClient.prototype.getRoom = function(roomId) {
...
+ * in case lazy loading of memberships is in use.
+ * @param {string} roomId The room ID
+ */
+MatrixClient.prototype.loadRoomMembersIfNeeded = function(roomId) {
+    const room = this.getRoom(roomId);
+    if (!room || !room.membersNeedLoading()) {
+        return;
+    }
+    const membersPromise = this.joinedMembers(roomId);
+    room.setLazilyLoadedMembers(membersPromise);
```
[the remark this answers is not in the document]
Comment (bwindels) [reply src/client.js]: Because putting the flag (`_membersNeedLoading`) whether loading has started in Room felt natural as it needs to be kept per room. I agree it's a bit weird, but modifying the flag, private field on Room, from MatrixClient would also not be pretty IMO. I didn't put the whole thing in Room as the pattern seems to be to do network I/O calls in client only. Wdyt?

I could still return a promise from here though, and put the error logging in RoomView then. I'll do that.

[the remark this answers is not in the document]
Comment (bwindels) [reply src/client.js]: @dbkr To clarify how the flag is used to avoid a race, I commented here in the follow-up PR. Makes more sense now?:
https://github.com/matrix-org/matrix-js-sdk/pull/669/commits/6570f56538be8198e69ff4308b69f11c47511fb6

**Comment by bwindels** [comment]
@dbkr I think the LL members should be set on every state of every timeline of every set (not sure about search, but don't see any reason why that's different?), as we can't know since when that LL member is valid, but you have a point the the LL members should never take precedence over members known through state, as this would indeed override historical data. I'll make changes so this is the case.

Concerning tests: yes, I'll write some :+1:

**Comment by bwindels** [comment]
Working on the tests, don't merge yet.

**Comment by dbkr** [comment]
Just been having a think about this. I think you're right in that we would need to set the members on every timeline in order to get members for events in previous timeline when scrolling back? However, this surely won't work if the member has since left the room since we won't get them in `/joinedMembers` anyway, so it seems like this is never going to work reliably without server support for back-paginating (this is exactly the sort of edge case a unit test would be perfect for, ie. back-paginating to an event from a member who has now left).

I think that's the only reason for setting members on anything but the forward edge of the live timeline? In general I would definitely rather avoid putting member events on historical bits of the timeline - especially when we might want to use this data to figure out who was in a room at the time a message was sent to decide whether to send e2e keys to someone.

We should also comment the corners we're cutting here, ie. I'd expect a massive XXX comment saying that `/jojnedMembers` lacks a way to get the members at a specific point and the resulting raciness.

**Comment by bwindels** [comment]
@dbkr Right, the only edge case I can think off is when you'd back paginate before the LL members are loaded, but that's probably avoidable.

Adding support for lazy loading in back-pagination is both the plan for client and server before releasing this, see https://github.com/vector-im/riot-web/issues/6611 (adding this to the description here as well), so you're right we don't really need to correct any other room state but the forward, live one. I'll change it back.

Added a big XXX corners cut comment where we call /joined_members :+1:

**Comment by bwindels** [comment]
@dbkr @t3chguy ready for another look

**Comment by dbkr** [comment]
Yeah, this feels a lot safer. Once we have a `/members` API that can get state for any given point, it might make sense to track the `membersNeedLoading` per roomState so we can also load member lists for these historical states. This can come later though.

**Comment by bwindels** [comment]
Just had a look at the /members endpoint, and as it returns state events, not just profile information, some of the changes here could probably be avoided. So I'll make the change as part of this PR.
